pkg/cluster/cluster.go

```diff
@@ -67,3 +71,111 @@ func (ln *localNode) Observe(ckit.Observer) {
 func (ln *localNode) Peers() []peer.Peer {
 	return []peer.Peer{ln.self}
 }
+
+// Clusterer implements the behavior required for operating Flow controllers
+// in a distributed fashion.
+type Clusterer struct {
+	Node Node
+	Mux  *http.ServeMux
+
+	RegisteredComponents map[string]struct{}
+}
+
+// New creates a Clusterer.
+func New(log log.Logger, clusterEnabled bool, addr, joinAddr string) (*Clusterer, error) {
+	// Standalone node.
+	if !clusterEnabled {
+		mux := http.NewServeMux()
+		mux.HandleFunc("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Write([]byte("clustering is disabled"))
+			w.WriteHeader(http.StatusBadRequest)
+		}))
+		return &Clusterer{
+			Node: NewLocalNode(addr),
+			Mux:  mux,
+
+			RegisteredComponents: make(map[string]struct{}),
+		}, nil
+	}
+
+	host, portStr, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, err
+	}
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return nil, err
+	}
+
+	mux := http.NewServeMux()
+
+	gossipConfig := DefaultGossipConfig
+	gossipConfig.NodeName = uuid.NewString()
+	gossipConfig.AdvertiseAddr = host
+	gossipConfig.ApplyDefaults(port)
+
+	if joinAddr != "" {
+		gossipConfig.JoinPeers = strings.Split(joinAddr, ",")
+	}
+
+	gossipNode, err := NewGossipNode(log, mux, &gossipConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	// Attempt to start the Node by connecting to the peers in gossipConfig.
+	// If we cannot connect to any peers, fall back to bootstrapping a new
+	// cluster by ourselves.
+	err = gossipNode.Start()
+	if err != nil {
+		level.Debug(log).Log("msg", "failed to connect to peers; bootstrapping a new cluster")
+		gossipConfig.JoinPeers = nil
+		err = gossipNode.Start()
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
This smells a bit hacky. We could delay the Start() call to after we start the HTTP server so we can always connect to ourselves without the need for special handling